src/DataFixtures/DownloadFixtures.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\DataFixtures;
+
+use App\Command\MigrateDownloadCountsCommand;
+use App\Entity\Package;
+use App\Entity\Version;
+use DateInterval;
+use DateTimeImmutable;
+use Doctrine\Bundle\FixturesBundle\Fixture;
+use Doctrine\Common\DataFixtures\DependentFixtureInterface;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\Persistence\ObjectManager;
+use Predis\Client;
+use Symfony\Component\Console\Helper\ProgressBar;
+use Symfony\Component\Console\Input\ArrayInput;
+use Symfony\Component\Console\Output\ConsoleOutput;
+
+/**
+ * Creates fake download statistics for each package.
+ */
+class DownloadFixtures extends Fixture implements DependentFixtureInterface
+{
+    private Client $redis;
+
+    private MigrateDownloadCountsCommand $migrateDownloadCountsCommand;
+
+    public function __construct(Client $redis, MigrateDownloadCountsCommand $command)
+    {
+        $this->redis = $redis;
+        $this->migrateDownloadCountsCommand = $command;
+    }
+
+    public function getDependencies()
+    {
+        return [
+            PackageFixtures::class,
+        ];
+    }
+
+    public function load(ObjectManager $manager)
+    {
+        $input  = new ArrayInput([]);
+        $output = new ConsoleOutput();
+
+        $output->writeln('Generating downloads...');
+
+        /** @var Package[] $packages */
+        $packages = $manager->getRepository(Package::class)->findAll();
+
+        $progressBar = new ProgressBar($output, count($packages));
+        $progressBar->setFormat('%current%/%max% [%bar%] %percent:3s%% (%remaining% left) %message%');
+
+        $progressBar->setMessage('');
+        $progressBar->start();
+
+        // Set the Redis keys that would normally be set by the DownloadManager, for the whole period.
+
+        foreach ($packages as $package) {
+            $progressBar->setMessage($package->getName());
+            $progressBar->display();
+
+            /** @var EntityManagerInterface $manager */
+            $latestVersion = $this->getLatestPackageVersion($manager, $package);
+
+            $this->populateDownloads($package, $latestVersion);
+
+            $progressBar->advance();
+        }
+
+        $progressBar->finish();
+        $output->writeln('');
+
+        // Then migrate the Redis keys to the db
+
+        $output->writeln('Migrating downloads to db... (this may take some time)');
+        $this->migrateDownloadCountsCommand->run($input, $output);
+    }
+
+    /**
+     * Returns the latest non-dev version of the given package.
+     */
+    private function getLatestPackageVersion(EntityManagerInterface $manager, Package $package): Version
+    {
+        return $manager->createQueryBuilder()
+            ->select('v')
+            ->from(Version::class, 'v')
+            ->where('v.package = :package')
+            ->setParameter('package', $package)
+            ->andWhere('v.normalizedVersion NOT LIKE :dev')
+            ->setParameter('dev', 'dev-%')
```

Comment:
You can check for `v.development = false` which is faster as it's backed by an index, and less error prone as this may return a version like 5.x-dev which I assume is not what you wanted.

Reply:
Thank you, updated.